fastcrypto-zkp/src/groth16/mod.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::fmt::Debug;
+
+use fastcrypto::error::{FastCryptoError, FastCryptoResult};
+use serde::Deserialize;
+
+use fastcrypto::groups::{GroupElement, Pairing};
+
+pub(crate) mod generic_api;
+
+#[derive(Debug, Deserialize)]
+pub struct Proof<G1: Pairing>
```

Comment:
while it's ok to how we use it today, isn't in general this is symmetric in the sense that a proof can be G2, G1, G2, where pairing is a trait of G1?

Reply:
I'm not sure I follow. You'd prefer a proof to be (G1, G2, G1) because it's smaller, assuming that we've always put G1 to be the smallest, right? But yes, the pairing is symmetric, so there's nothing stopping us from implementing `Pairing` as a trait for `G1::Other` for all `G1: Pairing`.